working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1949,31 +1986,27 @@ To match a pattern `p` against a value `v`:
 *   **Null-assert**:
 
     1.  If `v` is null then throw a runtime exception. *Note that we throw even
-        if this appears in a refutable context. The intent of this pattern is to
+        if this appears in a matching context. The intent of this pattern is to
         assert that a value *must* not be null.*
 
     2.  Otherwise, match the inner pattern against `v`.
 
-*   **Literal** or **constant**: The pattern matches if `o == v` evaluates to
-    `true` where `o` is the pattern's value.
-
-    **TODO: Should this be `v == o`?**
-
-*   **Variable**:
+*   **Constant**:
 
-    1.  If the runtime type of `v` is not a subtype of the static type of `p`
-        then the match fails.
+    1.  Evaluate the pattern's value to `c`.
 
-    2.  Otherwise, bind the variable's identifier to `v` and the match succeeds.
+    2.  The pattern matches if `c == v` evaluates to `true`. *This is opposite
+        the operand order that relational patterns use. This is deliberate to
+        preserve compatibility with existing switch cases and continue to enable
+        compilers to determine exactly which concrete `==` method is called in a
+        constant pattern for optimization purposes.*
 
-*   **Cast**:
+*   **Variable**:
 
-    1.  If the runtime type of `v` is not a subtype of the static type of `p`
-        then throw a runtime exception. *Note that we throw even if this appears
-        in a refutable context. The intent of this pattern is to assert that a
-        value *must* have some type.*
+    1.  If the runtime type of `v` is not a subtype of the static type of the
+        variable that `p` refers to then the match fails.
```

Comment:
This is an assignment, it can only happen in an irrefutable context.
So, instead of the match failing, we should throw a runtime error (and it will always be a failed downcast from `dynamic`).

Reply:
The sentence isn't clear, but the intent is that "the variable that `p` refers to" can also mean "the variable that `p` itself is declaring". So it can occur in a refutable context in which case it is a type test:

```dart
if ((123 as Object) case String s) { ... }
```

I'll reword this to make it clearer.